fs/parallel_artifact_fetcher.go

```diff
@@ -166,79 +157,87 @@ func (f *parallelArtifactFetcher) fetchFromRemoteAndWriteToTempDir(ctx context.C
 		return nil, fmt.Errorf("error truncating temp ingest file at %s: %w", ingestPath, err)
 	}
 
-	eg, egCtx := errgroup.WithContext(ctx)
-	var isSeekable atomic.Bool
-	isSeekable.Store(true)
+	doMultipleFetches := false
 	numLoops := f.calcNumLoops(desc.Size)
-
-	for i := range numLoops {
-		if !isSeekable.Load() {
-			// This val only gets set after the whole readcloser has been
-			// written to file, so safe to stop attempting this again.
-			break
+	if numLoops > 1 {
+		if rs, ok := f.remoteStore.(*orasBlobStore); ok {
+			// If this layer does not support ranged GET, it is very likely
+			// all other layers of this image do not either.
+			doMultipleFetches, err = rs.doInitialFetch(ctx, f.constructRef(desc))
+			if err != nil {
+				return nil, fmt.Errorf("error doing initial authorization for layer: %w", err)
+			}
 		}
+	}
+	if doMultipleFetches {
+		err = f.multiRequestFetchWrite(ctx, desc, file)
+	} else {
+		err = f.oneRequestFetchWrite(ctx, desc, file)
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("error writing to temp ingest file at %s: %w", ingestPath, err)
+	}
+
+	n, err := file.Seek(0, io.SeekStart)
+	if n != 0 {
+		err = errors.Join(err, errors.New("seek position != zero"))
+	}
+	if err != nil {
+		return nil, fmt.Errorf("error reopening file at %s after writing: %w", ingestPath, err)
+	}
+	return file, nil
+}
+
+// oneRequestFetchWrite does a normal fetch for the content from the repo and writes it to the given file descriptor
+func (f *parallelArtifactFetcher) oneRequestFetchWrite(ctx context.Context, desc ocispec.Descriptor, file *os.File) error {
+	err := f.layerUnpackJob.AcquireDownload(ctx, 1)
+	if err != nil {
+		return fmt.Errorf("error acquiring semaphore: %w", err)
+	}
+	defer f.layerUnpackJob.ReleaseDownload(1)
+
+	rc, err := f.remoteStore.Fetch(ctx, desc)
+	if err != nil {
+		return fmt.Errorf("error fetching from remote: %v", err)
+	}
+	return writeToEntireFile(file, rc)
+}
 
+// multiRequestFetchWrite will make parallel calls to the upstream repo for chunks of the file and buffer it from the given file descriptor.
+// This assumes that we will be fetching multiple chunks of the file and that we are using our own blob store implementation.
+func (f *parallelArtifactFetcher) multiRequestFetchWrite(ctx context.Context, desc ocispec.Descriptor, file *os.File) error {
+	eg, egCtx := errgroup.WithContext(ctx)
+	reference := f.constructRef(desc)
+
+	blobStore, ok := f.remoteStore.(*orasBlobStore)
+	if !ok {
+		return errors.New("did not pass orasBlobStore type to parallel fetcher")
+	}
+
+	for i := range f.calcNumLoops(desc.Size) {
```

Comment:
this is already calculated before. can be passed in as a parameter

Reply:
Sure makes sense, I didn't at first because it's part of the artifact fetcher but I guess it doesn't make sense to manually calculate this twice.